aten/src/ATen/native/CPUFallback.cpp

```diff
@@ -137,62 +166,124 @@ void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack, bool
   auto returns = torch::jit::last(stack, num_returns);
   const auto returns_begin = stack->size() - num_returns;
 
+  c10::optional<c10::Device> tgt_device =
+      compute_target_device(tensor_args, tensorlist_args);
+
   for (const auto idx : c10::irange(returns.size())) {
-    if (returns[idx].isTensor()) {
-      const auto& return_tens = returns[idx].toTensor();
-      if (return_tens.defined()) {
-        const AliasInfo* alias_info = schema_returns[idx].alias_info();
-        if (alias_info != nullptr && alias_info->isWrite()) {
-          // Case (1): mutable alias case. Move the input ivalue directly onto the stack
-          // in place of the existing cpu output tensor.
-          bool found_alias = false;
-          // We could store some extra metadata on the function schema to avoid the loop here
-          // if we need to improve perf.
-          for (const auto i : c10::irange(tensor_args_indices.size())) {
-            auto input_tensor_idx = tensor_args_indices[i];
-            const auto& input_tensor = cpu_tensors[i];
-            const AliasInfo* input_alias_info = schema_args[input_tensor_idx].alias_info();
-            // Checked above; adding assert to guard against breakage of the below condition due to changing the above if test.
-            TORCH_INTERNAL_ASSERT_DEBUG_ONLY(alias_info != nullptr);
-            if (input_tensor.defined() && (alias_info == input_alias_info || (input_alias_info != nullptr && *alias_info == *input_alias_info))) {
-              // We've found the original input tensor that aliases with the current output.
-              // Wrap it in an IValue and put it directly on the stack.
-              (*stack)[returns_begin + idx] = c10::IValue(tensor_args[i]);
-              found_alias = true;
-              break;
-            }
+    const AliasInfo* alias_info = schema_returns[idx].alias_info();
+    if (alias_info != nullptr && alias_info->isWrite()) {
+      // Case (1): mutable alias case.
+      // Move the input ivalue directly onto the stack in place of
+      // the existing cpu output tensor.
+      bool found_alias = false;
+      if (returns[idx].isTensor() && returns[idx].toTensor().defined()) {
+        // We could store some extra metadata on the function schema to avoid
+        // the loop here if we need to improve perf.
+        for (const auto i : c10::irange(tensor_args_indices.size())) {
+          auto input_tensor_idx = tensor_args_indices[i];
+          const auto& input_tensor = cpu_tensors[i];
```

Comment:
should this be `cpu_tensors[input_tensor_idx]`, not `cpu_tensors[i]`?

Reply:
It would also be great if you can add a test somewhere exercising the cpu fallback on one of the foreach ops (e.g. with cuda tensors), and verifying that the output is correct. Somewhere in `test/cpp/api` would probably be a reasonable place to put it. Maybe... `test/cpp/api/optim.cpp` if there isn't a more suitable test file, since foreach ops are mainly meant for use in optimizers.

Reply:
> should this be `cpu_tensors[input_tensor_idx]`, not `cpu_tensors[i]`?


I don`t change any codes about ```Tensor``` and I thought this should still be ```cpu_tensor[i]```, because ```cpu_tensors``` and ```tensor_args_indices``` are one-to-one correspondence.

Reply:
> It would also be great if you can add a test somewhere exercising the cpu fallback on one of the foreach ops (e.g. with cuda tensors), and verifying that the output is correct. Somewhere in `test/cpp/api` would probably be a reasonable place to put it. Maybe... `test/cpp/api/optim.cpp` if there isn't a more suitable test file, since foreach ops are mainly meant for use in optimizers.

Regarding testing, I have a question here and need your help.

All the foreach operators implemented by cpu and cuda have also been implemented, so it may not be possible to test the fallback scenario by default.

1. Is there some mechanism to uninstall some kernels that cuda has registered in the dispatcher?
2. Are there other operators (non-foreach) that meet the fallback scenario?

Reply:
Hmm... One pretty good option might actually be the open registration tests, defined here:

python driver code ([link](https://github.com/pytorch/pytorch/blob/main/test/test_cpp_extensions_open_device_registration.py#L91))
C++ registration code ([link](https://github.com/pytorch/pytorch/blob/main/test/cpp_extensions/open_registration_extension.cpp))

Since that custom backend test doesn't have any registrations for foreach ops, then calling a foreach op on tensors on the custom device in those tests should hit the fallback. Does that sound ok?

Reply:
Thank you for your advice.

The new commit is ready for review, please take a look again.

Thank you :D